internal/openstack/cache.go

```diff
@@ -15,150 +15,85 @@
-	Invalidate()
 }
 
-func (client *OpenStackClient) NewPortCache(ttl time.Duration, tag string, useFloatingIPs bool) (*SimplePortCache, error) {
-	networkingclient, err := client.NewNetworkV2()
-	if err != nil {
-		return nil, err
-	}
-
-	return NewPortCache(networkingclient, ttl, tag, useFloatingIPs), nil
-}
-
-func NewPortCache(networkingclient *gophercloud.ServiceClient, ttl time.Duration, tag string, useFloatingIPs bool) *SimplePortCache {
-	return &SimplePortCache{
+func NewPortClient(networkingclient *gophercloud.ServiceClient, tag string, useFloatingIPs bool) *UncachedClient {
+	return &UncachedClient{
 		client:         networkingclient,
-		ttl:            ttl,
 		tag:            tag,
 		useFloatingIPs: useFloatingIPs,
 	}
 }
 
-func (pc *SimplePortCache) GetPorts() ([]*portsv2.Port, error) {
-	err := pc.refreshIfInvalid()
+func (pc *UncachedClient) GetPorts() (ports []*portsv2.Port, err error) {
+	err = portsv2.List(
+		pc.client,
+		portsv2.ListOpts{Tags: pc.tag},
+	).EachPage(func(page pagination.Page) (bool, error) {
+		fetched_ports, err := portsv2.ExtractPorts(page)
+		if err != nil {
+			return false, err
+		}
+		for _, found_port := range fetched_ports {
+			ports = append(ports, &found_port)
+		}
+		return true, nil
+	})
 	if err != nil {
 		return nil, err
 	}
-
-	result := make([]*portsv2.Port, len(pc.ports))
-	i := 0
-	for _, v := range pc.ports {
-		result[i] = &v.Port
-		i += 1
-	}
-	return result, nil
+	return ports, err
 }
 
-func (pc *SimplePortCache) GetPortByID(ID string) (*portsv2.Port, *floatingipsv2.FloatingIP, error) {
-	err := pc.refreshIfInvalid()
+func (pc *UncachedClient) GetPortByID(ID string) (port *portsv2.Port, fip *floatingipsv2.FloatingIP, err error) {
+	port, err = portsv2.Get(
+		pc.client,
+		ID,
+	).Extract()
 	if err != nil {
 		return nil, nil, err
 	}
 
-	port, ok := pc.ports[ID]
-	if !ok {
-		return nil, nil, nil
-	}
-	return &port.Port, port.FloatingIP, nil
-}
-
-func (pc *SimplePortCache) refreshIfInvalid() error {
-	now := time.Now()
-	if now.Before(pc.validUntil) {
-		klog.V(5).Infof("not refreshing port cache, it is still valid until %s", pc.validUntil)
-		return nil
-	}
-	return pc.forceRefresh()
-}
-
-func (pc *SimplePortCache) forceRefresh() error {
-	// port ID -> FloatingIP, *NOT* floating IP ID -> FloatingIP
-	var fipEntries map[string]*floatingipsv2.FloatingIP = nil
 	if pc.useFloatingIPs {
-		fipEntries = make(map[string]*floatingipsv2.FloatingIP)
-		err := floatingipsv2.List(
+		err = floatingipsv2.List(
 			pc.client,
-			floatingipsv2.ListOpts{Tags: pc.tag},
+			floatingipsv2.ListOpts{Tags: pc.tag, PortID: ID},
 		).EachPage(func(page pagination.Page) (bool, error) {
 			fips, err := floatingipsv2.ExtractFloatingIPs(page)
 			if err != nil {
 				return false, err
 			}
-			for _, fip := range fips {
-				if fip.PortID == "" {
-					continue
+			for _, found_fip := range fips {
+				if fip != nil {
+					// TODO: warn here?!
```

Comment:
If this isn't supposed to happen warning here would be good (as it may help with debugging).